### ISSUE TITLE
`clang` and `gcc-11` failed to pass make -C tests test-lz4c32

### ISSUE THREAD

**Issue by t-mat** [lz4/lz4#991]
**Describe the bug**
When `CC=clang`, `make -C tests test-lz4c32` fails.
  - edit: Also fails with `CC=gcc-11` .

**Expected behavior**
When I use `CC=clang`, it passes all tests in `make -C tests test-lz4c32` without error.

**To Reproduce**

clone
```
cd
git clone --depth 1 --single-branch --branch dev https://github.com/lz4/lz4.git
cd lz4
git branch --show-current
# dev
git show -s --pretty=reference
# 634fa20 (Merge pull request #989 from lz4/fix_nullub_lz4, 2021-05-28)
```

Test with `gcc-10`

```
sudo apt-get install -y lib32gcc-10-dev
CC=gcc-10 make V=1 -C tests clean test-lz4c32
```

`gcc-10` passes all tests

```
# ...
# Files tmphf1 and tmphf2 are identical
# make: Leaving directory '/home/wsluser/lz4/tests'
```

Test with `clang-10`

```
CC=clang-10 make V=1 -C tests clean test-lz4c32
```

Test is aborted with the following assertion.

```
 ---- test huge files compression/decompression ----
./datagen -g6GB    | ../programs/lz4 -vB5D  | ../programs/lz4 -qt
using blocks of size 256 KB
*** LZ4 command line interface 32-bits v1.9.3, by Yann Collet ***
Read : 6126 MB   ==> 59.22%
Compressed 6442450944 bytes into 3815071561 bytes ==> 59.22%
./datagen -g4500MB | ../programs/lz4 -v3BD | ../programs/lz4 -qt
*** LZ4 command line interface 32-bits v1.9.3, by Yann Collet ***
Read : 2040 MB   ==> 45.27%   lz4: ../lib/lz4hc.c:282: int LZ4HC_InsertAndGetWiderMatch(LZ4HC_CCtx_internal *, const BYTE *const, const BYTE *const, const BYTE *const, int, const BYTE **, const BYTE **, const int, const int, const int, const dictCtx_directive, const HCfavor_e): Assertion `matchPtr >= lowPrefixPtr' failed.
Aborted
Error 68 : Unfinished stream
make: *** [Makefile:412: test-lz4-hugefile] Error 68
make: Leaving directory '/home/wsluser/lz4/tests'
```

Test with `gcc-11`

```
CC=gcc-11 make V=1 -C tests clean test-lz4c32
```

Test is aborted with the following assertion.

```
 ---- test huge files compression/decompression ----
./datagen -g6GB    | ../programs/lz4 -vB5D  | ../programs/lz4 -qt
using blocks of size 256 KB
*** LZ4 command line interface 32-bits v1.9.3, by Yann Collet ***
Read : 6095 MB   ==> 59.21%
Compressed 6442450944 bytes into 3814406542 bytes ==> 59.21%
./datagen -g4500MB | ../programs/lz4 -v3BD | ../programs/lz4 -qt
*** LZ4 command line interface 32-bits v1.9.3, by Yann Collet ***
Read : 2040 MB   ==> 45.26%   lz4: ../lib/lz4hc.c:148: LZ4HC_countBack: Assertion `(size_t)(match - mMin) < (1U<<31)' failed.
Error 68 : Unfinished stream
Aborted
make: *** [Makefile:412: test-lz4-hugefile] Error 68
make: Leaving directory '/home/wsluser/my-lz4/lz4/tests'
```

**System (please complete the following information):**
 - OS: Linux
 - Version: Ubuntu 20.04
 - Compiler
   - gcc-10 (Ubuntu 10.2.0-5ubuntu1~20.04) 10.2.0
   - clang version 10.0.0-4ubuntu1
     - with `lib32gcc-10-dev` and `gcc-multilib`
   - gcc-11 (Ubuntu 11.1.0-1ubuntu1~20.04) 11.1.0
     - with `lib32gcc-11-dev` and `gcc-multilib`
 - Build System: unmodified `Makefile` in the root directory of lz4.

**Additional context**
- When I execute same test with `-O0`, e.g. `CC=clang-10 CFLAGS=-O0 make ...`, this error doesn't happen. If I give `-O1`, `-O2` or `-O3`, same error happens.
  - `gcc-11` is similar.  `-O0` passes, `-O1`, `-O2`, `-O3` fails.
- Other versions of `clang` asserts same error.  Tested from `clang-3.5` to `clang-12`.  [Full logs](https://github.com/t-mat/lz4/runs/2703357090?check_suite_focus=true#step:8:1585).
- Honestly, I'm not confident that I'm using proper version of `lib32gcc` for specific version of `clang`.

**Comment by t-mat** [comment]
I observed same pattern of failure/success for x32.

```
CC=clang-10 CFLAGS='-O3 -mx32' make V=1 -C tests clean lz4 datagen test-lz4-hugefile
# -> lz4: ../lib/lz4hc.c:282: int LZ4HC_InsertAndGetWiderMatch(LZ4HC_CCtx_internal *, const BYTE *const, const BYTE *const, const BYTE *const, int, const BYTE **, const BYTE **, const int, const int, const int, const dictCtx_directive, const HCfavor_e)
#       : Assertion `matchPtr >= lowPrefixPtr' failed.
```

```
CC=clang-10 CFLAGS='-O0 -mx32' make V=1 -C tests clean lz4 datagen test-lz4-hugefile
# It passes all tests.
```

**Comment by t-mat** [comment]
I append about `gcc-11`.  But since it's fairly new compiler, potentially it's compiler/library/environment issue.

I observed that the following test fails with `gcc-11`.

```
CC=gcc-11 MOREFLAGS='-m32 -O3' make V=1 -C tests clean lz4 datagen test-lz4-hugefile
```

```
 ---- test huge files compression/decompression ----
...
# ./datagen -g4500MB | ../programs/lz4 -v3BD | ../programs/lz4 -qt
# Read : 2040 MB   ==> 45.26%   lz4: ../lib/lz4hc.c:148: LZ4HC_countBack: Assertion `(size_t)(match - mMin) < (1U<<31)' failed.
```

- `-O2` and `-O1` also fail the test.  Shows same assertion.

`-O0` passes it.

```
CC=gcc-11 MOREFLAGS='-m32 -O0' make V=1 -C tests clean lz4 datagen test-lz4-hugefile
# pass
```

**Comment by Cyan4973** [comment]
I confirm the issue, after reproducing it on a recent Ubuntu VM image with `gcc-11.1.0`

**Comment by Cyan4973** [comment]
strange heisenbug ....

When adding a few traces, I can get it to fail reliably, with following information : 
```
warning : (size_t)(match - mMin) >= (1U<<31) 
match : 0xf7cbe256 
mMin : 0xf78ce010 
lz4: ../lib/lz4hc.c:158: LZ4HC_countBack: Assertion `(size_t)(match - mMin) < (1U<<31)' failed.
Aborted (core dumped)
```

That's strange, because both pointers belong to the same 16 MB slice, starting with `0xf7...`, so their distance should not be that large to fail the following distance test.

Now, let's just add one more trace, displaying the resulting distance `(size_t)(match - mMin)` before the failing `assert()`.
When I do that ... the bug disappears, and test completes successfully.
Attempted many times, using different variant of calculating the distance, reliably reproduced every time.

That's weird, to say the least...

_edit_ : I could trick the trace into printing what is effectively the negative of the distance, in a way which makes the compiler think it's calculating something different from the following `assert()` test.
While pointer values change at each test, the distance is always the same, 4129350 bytes.
This is of course well short of the 2 GB being checked at next `assert()` test.

I'm currently suspecting a `gcc-11` "optimization", that somehow assumes the result of this `assert()` test without really calculating it.

**Comment by t-mat** [comment]
I think the following workaround works with `clang-[11,14]`.

```
// lz4hc.c
LZ4_FORCE_INLINE int LZ4HC_countBack(...) {
    ...
-   assert(match >= mMin);     assert((size_t)(match - mMin) < (1U<<31));
+   assert(match - mMin >= 0); assert((size_t)(match - mMin) < (1U<<31));
    ...
}

LZ4_FORCE_INLINE int LZ4HC_InsertAndGetWiderMatch(...) {
    ...
-   assert(matchPtr >= lowPrefixPtr);
+   assert(matchPtr - lowPrefixPtr >= 0);
    ...
}
```

But it worsens the situation with `gcc-[11,12]`.

Results of test-lz4c32

|Compiler		|without workaround	|with workaround	|
|--			|--			|--			|
|gcc-11			|Fail (lz4hc.c:148)	|Fail (lz4hc.c:282)	|
|gcc-12 experimental	|Pass			|Fail (lz4hc.c:282)	|
|clang-11		|Fail (lz4hc.c:282)	|Pass			|
|clang-12		|Fail (lz4hc.c:282)	|Pass			|
|clang-13		|Fail (lz4hc.c:282)	|Pass			|
|clang-14		|Fail (lz4hc.c:282)	|Pass			|

Compilers

|Compiler	|Version					|
|--		|--						|
|gcc-11		|gcc-11 (Ubuntu 11.2.0-19ubuntu1) 11.2.0	|
|gcc-12		|gcc-12 (Ubuntu 12-20220319-1ubuntu1) 12.0.1 20220319 (experimental) [master r12-7719-g8ca61ad148f]	|
|clang-11	|Ubuntu clang version 11.1.0-6			|
|clang-12	|Ubuntu clang version 12.0.1-19ubuntu3		|
|clang-13	|Ubuntu clang version 13.0.1-2ubuntu2		|
|clang-14	|Ubuntu clang version 14.0.0-1ubuntu1		|

I think this is a fundamental bug of optimizers of these compilers.  Some optimization for pointer arithmetics in `-m32` have serious issue.

**Comment by Cyan4973** [comment]
Going back to this issue, testing with `clang-12`,
I observe the reported issue `lz4hc.c:282`
immediately after the compression interface streams > 2 GB.

So this could be related to some kind of `signed` /  `unsigned` confusion in 32-bit mode.

As before, adding traces makes the issue disappear.

**Comment by Cyan4973** [comment]
> As before, adding traces makes the issue disappear.

Hey, maybe that could be a solution to this issue ?

_edit_ : nope, because traces work for `DEBUGLEVEL=2`, not `DEBUGLEVEL=1` (`assert()` only) which is the default for tests

**Comment by Cyan4973** [comment]
A fix is ongoing (#1112).

Question : was the issue discovered while trying to add more recent compilers to the CI test suite ?

**Comment by t-mat** [comment]
> Question : was the issue discovered while trying to add more recent compilers to the CI test suite ?

Yes.  This issue was discovered during trying to "complete" our compiler matrix (compiler vendors, versions, architectures).
